design-proposals/monitoring-code-refactor/monitoring-code-refactor.md

```diff
@@ -0,0 +1,112 @@
+# Monitoring code refactor
+
+**Authors:** João Vilaça <jvilaca@redhat.com>
+
+## What
+
+This design document is proposing a code refactor for the monitoring logic in
+all KubeVirt components. The goal is to have a centralized place for the
+creation of monitoring resources, a consistent monitoring package between all
+projects, a code struct that is easy to maintain and evolve, while moving closer
+to the Kubernetes metric implementation style.
+
+## Why
+
+It’s important to refactor the monitoring code to separate monitoring logic from
+business logic. It will make the codebase more modular, readable, and
+maintainable. It will reduce complexity, reduce the risk of introducing errors,
+and allow us to test each component independently.
+
+Right now, every change regarding monitoring needs to happen at many different
+places. For example, when adding the linter for metrics names, it was necessary
+to go to all KubeVirt components repositories and implement the linter
+validation in each. This is demanding, very time-consuming, requires going
+through multiple duplicated code reviews, uses many CI resources, and makes it
+mentally harder to decide to change anything.
+
+### Pitfalls of the current solution
+
+We are facing multiple challenges with the current solution:
+
+- It is hard to keep a mental model of how monitoring is implemented in all the
+different repositories and is difficult to maintain all of them in sync with new
+features
+
+- Hard to add new features to metrics and alerts, for example, because of how
+metrics are being created it is hard to add a new metric stability field to all
+of them since they are independently created in many different places
+
+- Complex to collect all existing metrics and alerts for automatic documentation
+
+- High levels of intertwined monitoring and business logic code
+
+### Goals
+
+- Decouple monitoring logic from business logic
+- Have a centralized place for the configuration of monitoring resources
+- Encapsulate the monitoring best-practices and the common patterns into a
+library and have it as a dependency for all KubeVirt components
+- Have a declarative way to define metrics and alerts
+- Keep monitoring code and utilities easy to maintain and evolve
+- Document how to use the monitoring library for adding new metrics and rules
+- Have a structure and tools to accurately and easily generate monitoring
+documentation, lint metrics and alerts, define allow-list, deny-list and
+opt-in metrics and other future features without having to change the code in
```

Comment:
Can you explain how it would be possible to opt in/out of metrics/alerts?

Reply:
the `operator-observability` package will already provide the idea of a `group` of metrics which will help with this.

We will have 2 flags:

opt-in: a list of groups and/or metrics that should be the only ones registered in the Prometheus Registry
opt-out: a list of groups and/or metrics that should be excluded from the registration in the Prometheus Registry (while everything else is registered)